src/main/java/picocli/BashCompletionAutoComplete.java

```diff
@@ -0,0 +1,656 @@
+
+        StringBuilder buff = new StringBuilder(1024);
+        String optionNames = flagOptionNames + (!argOptionNames.isEmpty() ? " " + argOptionNames : "");
+        buff.append(format(HEADER, functionName, commands, optionNames));
+
+        List<OptionSpec> enumOptions = filter(commandSpec.options(), new EnumArgFilter());
+
+        for (OptionSpec f : enumOptions) {
+            generateEnumCompetionCandidates(buff, f);
+        }
+
+        List<OptionSpec> optionsWithCompletionCandidates = filter(commandSpec.options(), and(new HasCompletions(), negate(new EnumArgFilter())));
+
+        for (OptionSpec f : optionsWithCompletionCandidates) {
+            generateNonEnumCompetionCandidates(buff, f);
+        }
+
+        for (PositionalParamSpec f : enumPositionalParams) {
+            generateEnumCompetionCandidates(buff, f);
+        }
+
+        List<PositionalParamSpec> positionalParametersWithCompletionCandidates = filter(commandSpec.positionalParameters(), and(new HasCompletions(), negate(new EnumArgFilter())));
+
+        for (PositionalParamSpec f : positionalParametersWithCompletionCandidates) {
+            generateNonEnumCompetionCandidates(buff, f);
+        }
+
+        buff.append(generateOptionsSwitch(argOptionFields, enumOptions));
+
+        buff.append(generateSubcommandsSwitch(commandLine, functionName));
+
+        buff.append(format(FOOTER));
+
+        return buff.toString();
+    }
+
+    private static Object generateSubcommandsSwitch(CommandLine commandLine, String functionName) {
+        StringBuilder buff = new StringBuilder(1024);
+
+        Set<String> subcommands = commandLine.getSubcommands().keySet();
+
+        if (subcommands.isEmpty()) {
+            return buff.toString();
+        }
+
+        buff.append("\n");
+        buff.append("    local arg\n");
+        buff.append("    _get_first_arg\n");
+        buff.append("\n");
+        buff.append("    case $arg in\n");
+        for (String subcommand : subcommands) {
+            buff.append("        ");
+            buff.append(subcommand);
+            buff.append(")\n");
+            buff.append("            ");
+            buff.append(functionName);
+            buff.append("_");
+            buff.append(subcommand);
+            buff.append("\n");
+            buff.append("            return\n");
+            buff.append("            ;;\n");
+        }
+        buff.append("    esac\n");
+
+        return buff.toString();
+    }
+    private static void generateEnumCompetionCandidates(StringBuilder buff, ArgSpec f) {
+        if (f instanceof OptionSpec) {
+            buff.append("\n");
+            buff.append(format("    local %s_option_args\n", bashify(f.paramLabel().toLowerCase(ENGLISH))));
+            buff.append(format("    %s_option_args=\'%s\'\n",
+                    bashify(f.paramLabel().toLowerCase(ENGLISH)),
+                    concat(" ", Arrays.asList((Enum[]) f.type().getEnumConstants()), null, new EnumNameFunction()).trim()));
```

Comment:
Doesn't `f.completionCandidates()` work also for ArgSpecs with an `enum` type? I thought that was fixed (so you should be able to combine `generateEnumCompetionCandidates` and `generateNonEnumCompetionCandidates` in one method).

Reply:
I've removed it, although this code was copied from the original generator.